File: app/src/main/java/com/mirash/passkeeper/view/gesturedetect/GestureDetectingScrollView.java
package com.mirash.passkeeper.view.gesturedetect;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.widget.ScrollView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mirash.passkeeper.view.gesturedetect.tool.GestureEventDetector;
import com.mirash.passkeeper.view.gesturedetect.tool.SingleTapUpListener;

/**
 * @author Mirash
 */
public class GestureDetectingScrollView extends ScrollView {
    private GestureEventDetector gestureDetector;

    public GestureDetectingScrollView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public GestureDetectingScrollView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    protected void init() {
        gestureDetector = new GestureEventDetector(getContext());
    }

    public void setSingleTapListener(SingleTapUpListener singleTapListener) {
        gestureDetector.setSingleTapListener(singleTapListener);
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        boolean scrollResult = super.onTouchEvent(ev);
        boolean gestureResult = gestureDetector.onTouchEvent(ev);
        return scrollResult || gestureResult;
    }
}

File: app/src/main/java/com/mirash/familiar/view/gesturedetect/GestureDetectingScrollView.kt
package com.mirash.familiar.view.gesturedetect

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.ScrollView
import com.mirash.familiar.view.gesturedetect.tool.GestureEventDetector
import com.mirash.familiar.view.gesturedetect.tool.SingleTapUpListener

/**
 * @author Mirash
 */
open class GestureDetectingScrollView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) :
    ScrollView(context, attrs, defStyleAttr) {
    private val gestureDetector: GestureEventDetector = GestureEventDetector(context)


    fun setSingleTapListener(singleTapListener: SingleTapUpListener?) {
        gestureDetector.setSingleTapListener(singleTapListener)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(ev: MotionEvent): Boolean {
        return super.onTouchEvent(ev) or gestureDetector.onTouchEvent(ev)
    }
}
